trainer/trainer.py

```diff
@@ -991,12 +991,19 @@ def _optimize(
                         loss_dict["amp_scaler"] = scaler.get_scale()  # for logging
                     update_lr_scheduler = scale_prev <= scaler.get_scale()
         else:
+            if hasattr(self.model, "before_main_optimizer_step"):
+                # model dependent control for custom backward and optimizer operations
+                # before the main optimizer step
+                self.model.before_main_optimizer_step(loss_dict, optimizer)
```

Comment:
This line should use ```TrainerCallbacks``` in callbacks.py to be widely reachable.

Reply:
@erogol  Is it ok to pass in `loss_dict` and `optimizer` to a callback?

Reply:
@erogol pushed up the callback, please check